core/src/main/scala/akka/persistence/jdbc/state/scaladsl/DurableStateSequenceActor.scala

```diff
@@ -83,104 +108,122 @@ class DurableStateSequenceActor[A](stateStore: JdbcDurableStateStore[A], config:
-          // we must decide if we move the cursor forward
-          val newMax =
-            if ((currentMax + 1).until(currentElement).forall(givenUp.contains)) {
-              // we move the cursor forward when:
-              // 1) they have been detected as missing on previous iteration, it's time now to give up
-              // 2) current + 1 == currentElement (meaning no gap). Note that `forall` on an empty range always returns true
-              currentElement
-            } else currentMax
-
-          // we accumulate in newMissing the gaps we detect on each iteration
-          val newMissing =
-            if (previousElement + 1 == currentElement || newMax == currentElement) missing
-            else missing.addRange(previousElement + 1, currentElement)
-
-          (newMax, currentElement, newMissing)
-      }
 
-    val newMissingByCounter = missingByCounter + (moduloCounter -> missingElems)
-
-    // did we detect gaps in the current batch?
-    val noGapsFound = missingElems.isEmpty
+    val (missingOffsetCurrentCount, nextMax, cacheMissed) = findRequeryCriteria(elements, currentMaxGlobalOffset)
 
     // full batch means that we retrieved as much elements as the batchSize
     // that happens when we are not yet at the end of the stream
     val isFullBatch = elements.size == batchSize
 
-    if (noGapsFound && isFullBatch) {
-      // Many elements have been retrieved but none are missing
+    if (missingOffsetCount + missingOffsetCurrentCount == 0 && isFullBatch && !cacheMissed) {
+      // Many elements have been retrieved but none are missing (no gaps)
       // We can query again immediately, as this allows the actor to rapidly retrieve the real max ordering
-      self ! QueryOrderingIds
-      context.become(receive(nextMax, newMissingByCounter, moduloCounter))
+      self ! QueryState
+      context.become(receive(nextMax, missingOffsetCount + missingOffsetCurrentCount, moduloCounter))
     } else {
-      // either we detected gaps or we reached the end of stream (batch not full)
+      // 3 conditions where we need to delay and re-query
+      //
+      // 1. we detected gaps
+      // 2. we reached the end of stream (batch not full)
+      // 3. we encountered cache miss, i.e. found a new pid not present in the cache. This is because if the
+      //    revision of this newly encountered pid is > 1 we need to see the revision 1 (which may be very old)
+      //    in order to find if it's still relevant
       // in this case we want to keep querying but not immediately
       scheduleQuery(queryDelay)
-      context.become(receive(nextMax, newMissingByCounter, (moduloCounter + 1) % maxTries))
+      context.become(receive(nextMax, missingOffsetCount + missingOffsetCurrentCount, (moduloCounter + 1) % maxTries))
+    }
+
+    // cache to be used in next iteration
+    if (!elements.isEmpty) {
+      elements.foreach(e => lastQueryResult += ((e.pid, e)))
```

Comment:
I think something is not optimal with updating this for each round, but I will have to play with some tests to see what happens.